grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/Fragments.java

```diff
@@ -174,12 +178,17 @@ public static Fragment resourceIndex(Var start, TypeLabel typeLabel, Object reso
         return traversal.union(__.not(__.has(INSTANCE_TYPE_ID.name())), __.repeat(__.in(SUB.getLabel())).emit()).unfold();
     }
 
-    static String displayOptionalTypeLabel(Optional<TypeLabel> typeLabel) {
-        return typeLabel.map(label -> " " + typeLabelToString(label)).orElse("");
+    static String displayOptionalTypeLabels(Optional<Set<TypeLabel>> typeLabels) {
+        return typeLabels.map(labels ->
+            " " + labels.stream().map(StringConverter::typeLabelToString).collect(joining(","))
+        ).orElse("");
     }
 
-    static void applyTypeLabelToTraversal(
-            GraphTraversal<Vertex, Edge> traversal, Schema.EdgeProperty property, Optional<TypeLabel> typeLabel, GraknGraph graph) {
-        typeLabel.ifPresent(label -> traversal.has(property.name(), graph.admin().convertToId(label).getValue()));
+    static void applyTypeLabelsToTraversal(
+            GraphTraversal<Vertex, Edge> traversal, Schema.EdgeProperty property, Optional<Set<TypeLabel>> typeLabels, GraknGraph graph) {
+        typeLabels.ifPresent(labels -> {
+            Set<Integer> typeIds = labels.stream().map(label -> graph.admin().convertToId(label).getValue()).collect(toSet());
+            traversal.has(property.name(), P.within(typeIds));
```

Comment:
Why does this need to take in multiple type labels now? I ask because I am not sure of the efficiency of using `P.within` with an indexed lookup. It should be fast but . . . you know.